src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs

```diff
@@ -1061,116 +1061,155 @@ public static unsafe SocketError GetLingerOption(SafeCloseSocket handle, out Lin
+            // only falling back to allocating it on the heap if it's deemed too big.
+
+            const int StackThreshold = 80; // arbitrary limit to avoid too much space on stack
+            if (count < StackThreshold)
+            {
+                Interop.Sys.PollEvent* eventsOnStack = stackalloc Interop.Sys.PollEvent[count];
+                return SelectViaPoll(
+                    checkRead, checkReadInitialCount,
+                    checkWrite, checkWriteInitialCount,
+                    checkError, checkErrorInitialCount,
+                    eventsOnStack, count, microseconds);
             }
-            finally
+            else
             {
-                if (releaseHandle)
+                var eventsOnHeap = new Interop.Sys.PollEvent[count];
+                fixed (Interop.Sys.PollEvent* eventsOnHeapPtr = eventsOnHeap)
                 {
-                    handle.DangerousRelease();
+                    return SelectViaPoll(
+                        checkRead, checkReadInitialCount,
+                        checkWrite, checkWriteInitialCount,
+                        checkError, checkErrorInitialCount,
+                        eventsOnHeapPtr, count, microseconds);
                 }
             }
-            return SocketError.Success;
         }
 
-        public static unsafe SocketError Select(IList checkRead, IList checkWrite, IList checkError, int microseconds)
+        private static unsafe SocketError SelectViaPoll(
+            IList checkRead, int checkReadInitialCount,
+            IList checkWrite, int checkWriteInitialCount,
+            IList checkError, int checkErrorInitialCount,
+            Interop.Sys.PollEvent* events, int eventsLength,
+            int microseconds)
         {
-            uint* readSet = stackalloc uint[Interop.Sys.FD_SETSIZE_UINTS];
-            int maxReadFd = Socket.FillFdSetFromSocketList(readSet, checkRead);
-
-            uint* writeSet = stackalloc uint[Interop.Sys.FD_SETSIZE_UINTS];
-            int maxWriteFd = Socket.FillFdSetFromSocketList(writeSet, checkWrite);
-
-            uint* errorSet = stackalloc uint[Interop.Sys.FD_SETSIZE_UINTS];
-            int maxErrorFd = Socket.FillFdSetFromSocketList(errorSet, checkError);
-
-            int fdCount = 0;
-            uint* readFds = null;
-            uint* writeFds = null;
-            uint* errorFds = null;
+            // Add each of the list's contents to the events array 
+            Debug.Assert(eventsLength == checkReadInitialCount + checkWriteInitialCount + checkErrorInitialCount, "Invalid eventsLength");
+            int offset = 0;
+            AddToPollArray(events, eventsLength, checkRead, ref offset, Interop.Sys.PollEvents.POLLIN | Interop.Sys.PollEvents.POLLHUP);
+            AddToPollArray(events, eventsLength, checkWrite, ref offset, Interop.Sys.PollEvents.POLLOUT);
+            AddToPollArray(events, eventsLength, checkError, ref offset, Interop.Sys.PollEvents.POLLPRI);
```

Comment:
>DangerousGetHandle is being used to get the underlying handle/file descriptor, without using DangerousAddRef/Release. 

The Sockets code is ~15 years old. And good practices such as always doing AddRef/Release while holding a bare handle (via `DangerouGetHandle` weren't clearly defined back then.  It's fine to leave as-is for now to match .NET Framework behavior. We should revisit. I don't think it will impact performance much to add those.